core/src/main/java/org/infinispan/util/WriteableCacheCollectionMapper.java

```diff
@@ -0,0 +1,96 @@
+package org.infinispan.util;
+
+import java.util.Collection;
+import java.util.function.Function;
+import java.util.function.Predicate;
+
+import org.infinispan.CacheCollection;
+import org.infinispan.commons.util.CloseableIterator;
+import org.infinispan.commons.util.CloseableIteratorMapper;
+import org.infinispan.commons.util.InjectiveFunction;
+
+/**
+ * A writeable cache collection mapper that also has constant time operations for things such as
+ * {@link Collection#contains(Object)} if the underlying Collection does.
+ * <p>
+ * This collection should be used for cases when a simple transformation of a element to another is all that is
+ * needed by the underlying collection.
+ * <p>
+ * Note this class allows for a different function specifically for values returned from an iterator. This
+ * can be useful to intercept calls such as {@link java.util.Map.Entry#setValue(Object)} and update appropriately.
+ * @author wburns
+ * @since 9.2
+ */
+public class WriteableCacheCollectionMapper<E, R> extends CacheCollectionMapper<E, R> {
+   protected final Function<? super E, ? extends R> toNewTypeIteratorFunction;
+   protected final Function<? super R, ? extends E> fromNewTypeFunction;
+
+   public WriteableCacheCollectionMapper(CacheCollection<E> realCollection,
+         Function<? super E, ? extends R> toNewTypeFunction,
+         Function<? super R, ? extends E> fromNewTypeFunction,
+         InjectiveFunction<Object, ?> keyFilterFunction) {
+      super(realCollection, toNewTypeFunction, keyFilterFunction);
+      this.toNewTypeIteratorFunction = toNewTypeFunction;
+      this.fromNewTypeFunction = fromNewTypeFunction;
+   }
+
+   public WriteableCacheCollectionMapper(CacheCollection<E> realCollection,
+         Function<? super E, ? extends R> toNewTypeFunction,
+         Function<? super E, ? extends R> toNewTypeIteratorFunction,
+         Function<? super R, ? extends E> fromNewTypeFunction,
+         InjectiveFunction<Object, ?> keyFilterFunction) {
+      super(realCollection, toNewTypeFunction, keyFilterFunction);
+      this.toNewTypeIteratorFunction = toNewTypeIteratorFunction;
+      this.fromNewTypeFunction = fromNewTypeFunction;
+   }
+
+   @Override
+   public CloseableIterator<R> iterator() {
+      return new CloseableIteratorMapper<>(realCollection.iterator(), toNewTypeIteratorFunction);
+   }
+
+   @Override
+   public boolean contains(Object o) {
```

Comment:
Unfortunately it can't implement it as it would need the function to convert from target type to storage type (ie. R to E). I can document this on `CacheCollectionMapper` as well as it currently is only on `CollectionMapper`

Reply:
Then maybe remove `CacheCollectionMapper`? I don't see it used anywhere else.

Reply:
Technically I can't remove it as it was there before. I also have to add back in the previous constructor without the `keyFilterMapper`.

Reply:
I will deprecate it though.

Reply:
There's no `@public` in the `org.infinispan.util.package-info`...